src/inmanta/scheduler/scheduler.py

```diff
@@ -0,0 +1,318 @@
+        print(f"Running {self._name}")
+
+    def priority(self) -> int:
+        return self._priority
+
+    def name(self) -> str:
+        return self._name
+
+    def __str__(self) -> str:
+        return self._name
+
+    def __repr__(self) -> str:
+        return self._name
+
+
+T = typing.TypeVar("T", bound="_typeshed.SupportsRichComparison")
+
+
+class PriorityQueue(asyncio.Queue[T]):
+    """
+    Completely sorted priority queue
+
+    less efficient than heapq
+    """
+
+    _queue: list[T]
+
+    def _init(self, maxsize: int) -> None:
+        self._queue = []
+
+    def _put(self, item: T) -> None:
+        bisect.insort(self._queue, item)
+
+    def _get(self) -> T:
+        return self._queue.pop(0)
+
+    def drop(self, item: T) -> None:
+        """Remove this item, raise indexerror if the item is not present"""
+        idx = bisect.bisect(self._queue, item) - 1
+        assert self._queue[idx] == item
+        self._queue.pop(idx)
+
+
+class TaskQueue:
+    """Queue to schedule tasks with inter-dependencies and provide an overview of the queue"""
+
+    def __init__(self) -> None:
+        self.full_queue: PriorityQueue[typing.Tuple[int, int, BaseTask]] = PriorityQueue()
+        self.run_queue: PriorityQueue[typing.Tuple[int, int, BaseTask]] = PriorityQueue()
+        self.counter = itertools.count()  # unique sequence count, makes sort stable in the heap
+
+    def put(self, task: BaseTask) -> None:
+        "Add a new task"
+        if task.cancelled:
+            return
+        count = next(self.counter)
+        entry = (task.priority(), count, task)
+        task._enqueue(self, entry)
+        self.full_queue.put_nowait(entry)
+        if task.waitcount == 0:
+            task._runnable()
+
+    async def get(self) -> BaseTask:
+        """Remove and return the lowest priority task. Raise IndexError if empty."""
+        priority, count, task = await self.run_queue.get()
+        return task
+
+    async def do_next(self) -> BaseTask:
+        """run the next task"""
+        task = await self.get()
+        await task._run()
+        return task
+
+    def view(self) -> typing.Sequence[Task]:
+        """This leaks a reference to the internal queue: don't touch it"""
+        return self.full_queue._queue
+
+
+class SentinelTask(BaseTask):
+    """Task to unblock the queue on shutdown"""
+
+    def name(self) -> str:
+        return "Queue Shutdown"
+
+    def priority(self) -> int:
+        return 0
+
+    async def run(self) -> None:
+        return
+
+
+class TaskRunner:
+    """Logical thread to process work from a task queue
+
+    Can be restarted
+    """
+
+    def __init__(self, queue: TaskQueue) -> None:
+        self.queue = queue
+        self.running = False
+        self.should_run = True
+        self.finished: typing.Optional[asyncio.Task[None]] = None
+
+    def start(self) -> None:
+        self.should_run = True
+        self.finished = asyncio.create_task(self.run())
+
+    def stop(self) -> None:
```

Comment:
This stops the runner asyncronously. I think it could use a docstring on what exactly it achieves.